### PR TITLE
WL 21727 - Music Visualizer Updates (marketplace item)

### PR DESCRIPTION
**Pull request by cain-kilgore**
> Current asset is here:
> http://mpassets.highfidelity.com/ced4e659-3a5a-442b-8de4-1acdf68e484e-v1/musicVisualizer.js
> Please update this script for better particle appearance (currently they're too fast, fuzzy, etc) and make sure it fully cleans up once script is closed. Resubmit to Marketplace when complete.
> [Worklist 21727](https://worklist.net/21727)

## Test Plan
1. Run the script from URL here: `http://mpassets.highfidelity.com/94c95207-e34c-48c8-afae-ad7dd9499a69-v1/musicVisualizer.js`
2. Confirm 'MUS VIS' gets added to your Toolbar/Tablet.
3. Supply a WAV file as the file and go through each effect.
4. Make sure there's no weird visual glitches and that it should look nice and smooth.
5. Click "Clean Up" in the app and confirm it deletes all existing Visualizers.
6. With visualizers in your sandbox, connect to a different domain then back to your sandbox.
7. Confirm that the Visualizers are still there, and can be cleared using the Clean Up button in the app.
8. Confirm that restarting the script also deletes the Visualizers.

example

### CODE REVIEW

**Comment by misslivirose** [comment]
Can the clean up button more closely align with the style of the other effect types?

File: Marketplace/musicVisualizer/adjuster_scripts/audioFileSync.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    // to enable gradual growth/decline of particle size
+    function mixValue(valueA, valueB, percentage) {
+        return valueA + ((valueB - valueA) * percentage);
+    }
+
+    _this.intervalID = Script.setInterval(function() {
+        if (injector !== undefined && injector.isPlaying()) {
+            // update spatialized sound location with particle movement
+            var entityPosition = Entities.getEntityProperties(_this.entityID, 'position').position;
+            if (Vec3.distance(lastPosition, entityPosition) > 0.01) {
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/adjuster_scripts/audioFileSync.js]: Magic number 0.01 should be named

File: Marketplace/musicVisualizer/adjuster_scripts/audioFileSync.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                    var effectProps = Entities.getEntityProperties(_this.entityID, 'particleRadius');
+                    effectProps.particleRadius = DEFAULT_RADIUS;
+                    lastUpdateLevel = 0;
+                    Entities.editEntity(_this.entityID, effectProps);
+                }
+            // update particle size with loudness of audio
+            } else {
+                var effectPropsChange = Entities.getEntityProperties(_this.entityID, 'particleRadius');
+                var targetLevel = injector.getLoudness() + DEFAULT_RADIUS;
+                effectPropsChange.particleRadius = mixValue(effectPropsChange.particleRadius, targetLevel, 0.5);
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/adjuster_scripts/audioFileSync.js]: Magic number

File: Marketplace/musicVisualizer/adjuster_scripts/micSync.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            if (lastUpdateLevel !== 0) {
+                var effectProps = Entities.getEntityProperties(_this.entityID, 'particleRadius');
+                effectProps.particleRadius = DEFAULT_RADIUS;
+                lastUpdateLevel = 0;
+                Entities.editEntity(_this.entityID, effectProps);
+            }
+        } else {
+            var effectPropsChange = Entities.getEntityProperties(_this.entityID, 'particleRadius');
+            var targetLevel = Audio.inputLevel + DEFAULT_RADIUS;
+            effectPropsChange.particleRadius = mixValue(effectPropsChange.particleRadius, targetLevel, 0.5);
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/adjuster_scripts/micSync.js]: Magic number 0.5

File: Marketplace/musicVisualizer/musVisLib.js [hunk 4]
```diff
@@ -0,0 +1,723 @@
+//  
+//  emojiLib.js
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musVisLib.js]: Update header with proper file naming and description

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+var AUDIO_SYNC_SCRIPT = Script.resolvePath("adjuster_scripts/audioFileSync.js?" + Date.now());
+var TRAIL_LEFT_SCRIPT = Script.resolvePath("adjuster_scripts/effectTrailerLeft.js?" + Date.now());
+var TRAIL_RIGHT_SCRIPT = Script.resolvePath("adjuster_scripts/effectTrailerRight.js?" + Date.now());
+
+var CLEAR_SELECTION_TEXT = "Clear selection";
+
+(function() {
+
+    var APP_NAME = "MUS VIS";
+    var APP_URL = "https://hifi-content.s3.amazonaws.com/elisalj/music_visualizer/musicVisualizerUI.html?" + Date.now();
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: We probably want this to be relative, so it's referencing the new HTML with cleanup button

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    tablet.screenChanged.connect(onScreenChanged);
+
+    // Gives position right in front of user's avatar
+    function getPositionToCreateEntity() {
+        var direction = Quat.getFront(MyAvatar.orientation);
+        var distance = 0.3;
+        var position = Vec3.sum(MyAvatar.position, Vec3.multiply(direction, distance));
+        position.y += 0.5;
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: Magic number 0.5

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        return position;
+    }
+
+    // adds the particle to world
+    function createParticle(effectName, micSync, behavior, file) {
+        print(effectName + " particle is being added");
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: I'd suggest removing the print statements in this file

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+//  Copyright High Fidelity 2017
+//  
+//  Licensed under the Apache 2.0 License
+//  See accompanying license file or http://apache.org/
+//  
+//  All assets are under CC Attribution Non-Commerical
+//  http://creativecommons.org/licenses/
+//  
+
+var LIB = Script.require("./musVisLib.js?" + Date.now());
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: Can this be renamed to something more descriptive / not abbreviated?

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+            var invisSphere = Entities.addEntity(invisible);
+            effectJSON.parentID = invisSphere;
+            var behaviorParticle = Entities.addEntity(effectJSON);
+            existingParticles.push(behaviorParticle);
+        }
+
+        // creates trail effect
+        if (behavior === "finger") {
+            effectJSON.isEmitting = false;
+            effectJSON.lifespan = 2;
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: magic number 2

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+                Entities.editEntity(effectRight, effectRightProps);
+
+                var equipLeftProps = Entities.getEntityProperties(equipLeft, "visible");
+                var equipRightProps = Entities.getEntityProperties(equipRight, "visible");
+                equipLeftProps.visible = true;
+                equipRightProps.visible = true;
+                Overlays.editOverlay(equipLeft, equipLeftProps);
+                Overlays.editOverlay(equipRight, equipRightProps);
+
+            }, 700);
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: Magic number 700

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+            audioFile = "";
+        }
+        tablet.emitScriptEvent(JSON.stringify({
+            "file": filename
+        }));
+    }
+
+    function onWebEventReceived(event) {
+        var htmlEvent = JSON.parse(event);
+        print("Event: " + JSON.stringify(htmlEvent));
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: Remove print statements

File: Marketplace/musicVisualizer/musicVisualizer.js [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        }
+    }
+
+    function cleanUpParticles() {
+        print("Cleaning up particles");
+        
+        var getNearbyEntities = Entities.findEntities(MyAvatar.position, DELETE_RADIUS);
+        for (var i = 0; i < getNearbyEntities.length; i++) {
+            var nearbyEntity = getNearbyEntities[i];
+            var nearbyEntityProperties = Entities.getEntityProperties(nearbyEntity);
```
Comment (misslivirose) [inline Marketplace/musicVisualizer/musicVisualizer.js]: Please only request the properties that you need when calling getEntityProperties

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Run the script from URL here: http://mpassets.highfidelity.com/94c95207-e34c-48c8-afae-ad7dd9499a69-v1/musicVisualizer.js **Check**
Confirm 'MUS VIS' gets added to your Toolbar/Tablet. **Check**
Supply a WAV file as the file and go through each effect. **Check**
Make sure there's no weird visual glitches and that it should look nice and smooth. **Check**
Click "Clean Up" in the app and confirm it deletes all existing Visualizers. **Check**
With visualizers in your sandbox, connect to a different domain then back to your sandbox. **Check**
Confirm that the Visualizers are still there, and can be cleared using the Clean Up button in the app. **Check**
Confirm that restarting the script also deletes the Visualizers. **Check**

All **Pass** @conklin94122

**Pull request merged**